### DISCUSSION TITLE
Retriggering Resources.

### DISCUSSION THREAD

**Discussion by squeedee** [vmware-tanzu/cartographer#657] [Ideas]
TL;DR - The problem with retriggering resources boils down to there being no good, accepted mechanism to do so. I highlight two workarounds below and then ask for contributions.

## stamp
Cartographer stamps resource yaml onto a cluster. It decides the shape of the resource yaml based on:
1. a template
2. paramarers
3. inputs.

## apply
Once applied (created or updated) the resource's own controller (outside of Carto) decides what to do with the information and, hopefully, reconciles with fields in its status that Carto treats as the artifact from the resource. This is it's _output_ 

## sources of failure that do not require edits
Several failures can occur when a resource is processing its `spec` that are not related to the quality of the inputs.
Eg. A network, disk or machine failure, out of memory errors and mid-operation interupts (scale/failover).

These kinds of issues should not require a user to intervene by changing an input (compared to, eg: when the code cannot compile)

A user simply want's the resource to recover.

## Repair scenarios for failed resources
1. The resource eventually self heals, leading to the correct, consistent output. They should all do this, but they wont necessarily. Especially those using Runnable
2. Changing the spec of the resource in someway triggers a reconcile that eventually leads to the correct, consistent output
3. Removing the resource, causing Carto to recreate the object, leads to the correct, consistent output
4. It's possible that some resources will base it's results on off-cluster resources, where the output can never be correct? I think this is highly unlikely.

## Workaround for scenario 2

If all we need is to change the resource's spec, we can use a parameter. Individual control could be provided with individualized params:

In the workload: 
```yaml
apiVersion: carto.run/v1alpha1
kind: Workload
metadata:
  name: #@ data.values.workload_name
  labels:
    app.tanzu.vmware.com/workload-type: web
spec:
  params:
    - name: rerunStamp
    - value: 2022-02-22T21:50:47Z. # any unique value would do.
```

in the template:
```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterSourceTemplate
metadata:
  name: source
spec:
  urlPath: .status.artifact.url
  revisionPath: .status.artifact.revision

  template:
    apiVersion: source.toolkit.fluxcd.io/v1beta1
    kind: GitRepository
    metadata:
      name: $(workload.metadata.name)$
      annotations:
        rerunStamp: $(params.rerunStamp)$ # this may cause the resource to re-evaluate?
    spec:
      interval: 1m0s
      url: $(workload.spec.source.git.url)$
      ref: $(workload.spec.source.git.ref)$
      gitImplementation: libgit2
      ignore: ""
```

**Cons with this approach**
* I'm confident this wont always be sufficient
* Requires that the retriggering params are baked into a supply chain.
* Would never consider this a formal API

## Workaround for scenario 3

If the only way a resource will try again (this I believe is true for any runnable) is to delete it, then that's the workaround. 

Simply delete the stamped resource and Cartographer will re-create it, hopefully causing the correct output to be created and triggering subsequent resources.

**Cons with this approach**
* Developer's often don't have access to do this. Supply chains should be inviolate
* Would never consider this a formal API

**Comment by squeedee** [comment]
I forgot to ask for contributions :D

It seems to me we can properly adress this by signaling to Cartographer that a resource needs to be deleted and re-created, but that leaves some open questions:
1. Is delete/recreate adequate in all/enough cases? (I think probably)
2. Does deleting the resource cause any issues with historical context of a workload? (I think probably not)
3. What's the clean, most 'k8s' way of signalling to the Carto Controller?
   1. A field on the workload
   2. A separate `control-resource` that selects for the workload and a resource (or whole supply chain) to rerun  
   3. A web interface directly on the controller allowing users to send it commands?
   4. Something far better that I haven't thought of?

Looking forward to some feedback :D
